### PR TITLE
Contribution guidelines document

### PR DESCRIPTION
**Pull request by fgalan**
Use the "View" buttom to see the rendered document. Check specially that all code blocks are ok and according to the surrounding rule.

~~In addition, the *How to check* section needs to be completed in each rule before merging this PR.~~

### CODE REVIEW

**Comment by kzangeli** [comment]
LGTM (I edited the 'big' file a little)

File: README.md [hunk 1]
```diff
@@ -145,6 +145,8 @@ Please have a look to the section [on building the source code](doc/manuals/admi
 * Container-based deployment
   * [Docker](docker/README.md)
 * [Sample code contributions](doc/manuals/code_contributions.md)
+* [Contribution guide](doc/manuals/contribution_guide.md), specially important if you plan to contribute with code
```
Comment (jmcanterafonseca) [inline README.md]: nit: 'especially' see 

http://dictionary.cambridge.org/es/gramatica/gramatica-britanica/especially-or-specially

Comment (jmcanterafonseca) [inline README.md]: better 'Contribution guidelines'

File: doc/manuals/contribution_guide.md [hunk 2]
```diff
@@ -0,0 +1,674 @@
+# Orion Context Broker Contribution Guide
+
+This document describes the guidelines to contribute to the Orion Context Broker open source repository. If you are
```
Comment (jmcanterafonseca) [inline doc/manuals/contribution_guide.md]: to contribute to Orion Context Broker . I think open source repository is not needed

File: doc/manuals/contribution_guide.md [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+well. This sometimes leads to conflicts and must be avoided. In addition, it increases the compilation time.
+
+*How to check*": manually?, make lint?, other? TBD
+
+#### M2 (Copyright header)
+
+*Rule*: Every file, source code or not, MUST have a copyright header:
+
+For C++ files:
+
```
Comment (jmcanterafonseca) [inline doc/manuals/contribution_guide.md]: perhaps we can have a template file and not display all this comment in this document

File: doc/manuals/contribution_guide.md [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+#
+# Author: <the author>
+```
```
Comment (jmcanterafonseca) [inline doc/manuals/contribution_guide.md]: same comment as above

File: doc/manuals/contribution_guide.md [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+*   About the components in a URL: according to 
+*   https://tools.ietf.org/html/rfc3986#section-3,
+*   the scheme component is mandatory, i.e. the 'http://' or 'https://' must 
+*   be present, otherwise the URL is invalid.
+*/
+```
+
+*Rationale*: the code is simply easier to read when prepared like this
+
+*How to check*": manually?, make lint?, other? TBD
```
Comment (jmcanterafonseca) [inline doc/manuals/contribution_guide.md]: if we don't know yet how to check, maybe we can leave out this part and open an issue on lintering

File: doc/manuals/contribution_guide.md [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+
+Some patterns are not allowed in Orion Context Broker code, except if some strong reason justifies the use of it. 
+If you plan to use some of them, please consult before with the core developers.
+
+* Multiple inheritance. It introduces complexity in the code. It is generally accepted that it is a “problematic”
+  pattern.
+* C++ exceptions 
+  * Exceptions (not checked in C++) make it harder to follow the flow of execution. It is hard to say whether
+  a function will throw an exception, complicating refactoring and debugging.
+  * Exception-safe functions requires assuring the transactional semantic is kept, (beyond RAII to avoid resource leaks),
```
Comment (jmcanterafonseca) [inline doc/manuals/contribution_guide.md]: nit: ensuring

File: doc/manuals/contribution_guide.md [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+  simplicity are allowed. Big savings in simplicity with small penalty in efficiency are also allowed.
+* New files contributed to Orion must follow the [filesystem layout guidelines](#filesystem-layout-guidelines).
+* Code contributed to Orion Context Broker must follow the [code style guidelines](#code-style-guidelines) 
+  in order to set a common programming style for all developers working on the code.
+* See also the section on [programming patterns](#programming-patterns) about which ones are allowed and disallowed in the 
+  Orion Context Broker code.
+
+Note that contribution workflows themselves (e.g. pull requests, etc.) are described in another document 
+([FIWARE Development Guide](https://forge.fiware.org/plugins/mediawiki/wiki/fiware/index.php/Developer_Guidelines)). 
+
```
Comment (jmcanterafonseca) [inline doc/manuals/contribution_guide.md]: FIWARE Development Guidelines

File: doc/manuals/contribution_guide.md (hunk 3 above)
Comment (fgalan) [inline doc/manuals/contribution_guide.md]: I though it that alternative but, at the end, given that the header fragment is not very long, I think is better to have everything in the same place.

File: doc/manuals/contribution_guide.md (hunk 4 above)
Comment (fgalan) [inline doc/manuals/contribution_guide.md]: Same answer :)

File: doc/manuals/contribution_guide.md (hunk 5 above)
Comment (fgalan) [inline doc/manuals/contribution_guide.md]: After @kzangeli 's last commit, all the rules has the "How to check" correctly filled.

Improving automatic procedures to become all the remainer "manually" to somekind of "checked automatically doing X" is left as future work  (maybe a good item for the bounty program?).

File: doc/manuals/contribution_guide.md (hunk 7 above)
Comment (fgalan) [inline doc/manuals/contribution_guide.md]: Fixed in ecddabee6e724137ad885176a9e9ce45d69f0f08

File: doc/manuals/contribution_guide.md (hunk 6 above)
Comment (fgalan) [inline doc/manuals/contribution_guide.md]: Fixed in ecddabee6e724137ad885176a9e9ce45d69f0f08

File: README.md (hunk 1 above)
Comment (fgalan) [inline README.md]: Fixed in ecddabee6e724137ad885176a9e9ce45d69f0f08

File: doc/manuals/contribution_guide.md (hunk 2 above)
Comment (fgalan) [inline doc/manuals/contribution_guide.md]: Fixed in ecddabee6e724137ad885176a9e9ce45d69f0f08

**Comment by jmcanterafonseca** [comment]
LGTM

good work!

**Comment by crbrox** [comment]
LGTM

**Pull request merged**